### PR TITLE
[Merged by Bors] - fix: ml-gateway (CT-000)

### PR DESCRIPTION
**Pull request by DecathectZero**
This finishes the general-runtime part of this refactor:
image

All instances of general-runtime DIRECTLY calling openAI or anthropic are now gone - it is all done through `ml-gateway`.

### CODE REVIEW

File: lib/services/runtime/handlers/utils/ai.ts [hunk 1]
```diff
@@ -52,78 +55,76 @@ export const fetchChat = async (
...
 
-  const workspaceID = runtime.project?.teamID;
-
-  if (typeof tokens === 'number' && tokens > 0) {
-    await runtime.services.billing
-      .trackUsage(ResourceType.WORKSPACE, workspaceID, ItemName.AITokens, tokens)
-      .catch((err: Error) =>
-        log.error(`[${reference}] Error consuming quota for workspace ${workspaceID}: ${log.vars({ err })}`)
-      );
-  }
```
Comment (DecathectZero) [inline lib/services/runtime/handlers/utils/ai.ts:122]: billing is now handled by `ml-gateway`

File: lib/services/runtime/handlers/aiResponse.ts [hunk 2]
```diff
@@ -23,8 +22,7 @@ const AIResponseHandler: HandlerFactory<VoiceNode.AIResponse.Node, void, General
     const nextID = node.nextId ?? null;
     const elseID = node.elseId ?? null;
     const projectID = runtime.project?._id;
-    const workspaceID = runtime.project?.teamID;
-    const generativeModel = runtime.services.ai.get(node.model);
+    const workspaceID = runtime.project?.teamID || '';
```
Comment (pmvrmc) [inline lib/services/runtime/handlers/aiResponse.ts:25]: should we error/warn if no workspaceID is present? it should always be there no?

File: lib/services/runtime/handlers/aiSet.ts [hunk 3]
```diff
@@ -15,9 +14,8 @@ const AISetHandler: HandlerFactory<BaseNode.AISet.Node, void, GeneralRuntime> =
   // eslint-disable-next-line sonarjs/cognitive-complexity
   handle: async (node, runtime, variables) => {
     const nextID = node.nextId ?? null;
-    const workspaceID = runtime.project?.teamID;
+    const workspaceID = runtime.project?.teamID || '';
```
Comment (pmvrmc) [inline lib/services/runtime/handlers/aiSet.ts:17]: same workspace comment as before, should it always exist?

File: lib/services/runtime/handlers/utils/generativeNoMatch.ts [hunk 4]
```diff
@@ -15,14 +13,11 @@ export const getCurrentTime = ({ newlines = 1 }: { newlines?: number } = {}) =>
 export const generateNoMatch = async (
   runtime: Runtime,
   context: BaseUtils.ai.AIModelParams
-): Promise<{ output: Output; tokens: number; queryTokens: number; answerTokens: number } | null> => {
+): Promise<AIResponse | null> => {
   if (context.model === BaseUtils.ai.GPT_MODEL.GPT_4 && runtime.plan && !GPT4_ABLE_PLAN.has(runtime.plan)) {
```
Comment (pmvrmc) [inline lib/services/runtime/handlers/utils/generativeNoMatch.ts:17]: eventually I think this check will go against billing too

File: lib/services/runtime/handlers/utils/generativeNoMatch.ts [hunk 5]
```diff
@@ -34,17 +29,17 @@ export const generateNoMatch = async (
     },
   ];
 
-  const model = runtime.services.ai.get(context.model);
-  const response = await fetchChat({ ...context, messages }, model, {
+  const response = await fetchChat({ ...context, messages }, runtime.services.mlGateway, {
     context: {
       projectID: runtime.project?._id,
-      workspaceID: runtime.project?.teamID,
+      workspaceID: runtime.project!.teamID,
```
Comment (pmvrmc) [inline lib/services/runtime/handlers/utils/generativeNoMatch.ts:35]: related to the previous message about workspaceID.

previously we default back to an empty string, in here we say it should always be there. We should make the same decision everywhere no?

File: lib/services/runtime/handlers/utils/knowledgeBase/index.ts [hunk 6]
```diff
@@ -204,7 +200,9 @@ export const knowledgeBaseNoMatch = async (
     if (faq?.answer) {
       addFaqTrace(runtime, faq.question || '', faq.answer, question.output);
       return {
-        output: generateOutput(faq.answer, runtime.project),
+        model: 'faq',
```
Comment (pmvrmc) [inline lib/services/runtime/handlers/utils/knowledgeBase/index.ts:203]: nit: move this faq to the same place as the other models? maybe `BaseUtils.ai.FAQ_MODEL.DEFAULT` ?

**Review by pmvrmc — APPROVED** [review]
looks good, some minor comments

File: lib/services/runtime/handlers/utils/knowledgeBase/index.ts (hunk 6 above)
Comment (DecathectZero) [reply lib/services/runtime/handlers/utils/knowledgeBase/index.ts:203]: I can actually just pass in the model of the question synthesis:
```
        model: question.model,
        multiplier: 1,
        output: faq.answer,
        tokens: question.queryTokens + question.answerTokens,
        queryTokens: question.queryTokens,
        answerTokens: question.answerTokens,
```

File: lib/services/runtime/handlers/aiResponse.ts (hunk 2 above)
Comment (DecathectZero) [reply lib/services/runtime/handlers/aiResponse.ts:25]: I'll do a separate refactor for that. You are right that it should always be there.
In the future the workspaceID will be more and more important.

**Pull request closed without merging**